File: java/src/main/java/br/com/emendes/problems/PartitionList.java
package br.com.emendes.problems;

import br.com.emendes.problems.util.ListNode;

/**
 * Dado o head de uma Linked List e um valor x, particione de maneira que todos os nodes menores que x
 * venham antes que os nodes maiores ou iguais a x.<br>
 * <br>
 * Você deve preservar a ordem relativa dos nodes em cada uma das partições.<br>
 * <br>
 * Retrições:<br>
 * O número de nodes na lista está no intervalo [0, 200].<br>
 * -100 <= Node.val <= 100<br>
 * -200 <= x <= 200
 */
public class PartitionList {

  public ListNode partition(ListNode head, int x) {
    if (x <= -100 || x > 100) return head;

    ListNode beforeDummy = new ListNode(0);
    ListNode afterDummy = new ListNode(0);
    ListNode before = beforeDummy;
    ListNode after = afterDummy;

    while (head != null) {
      if (head.val < x) {
        before.next = head;
        before = head;
      } else {
        after.next = head;
        after = head;
      }
      head = head.next;
    }
    before.next = afterDummy.next;
    after.next = null;

    return beforeDummy.next;
  }

}

File: javascript/problems/PartitionList.js
/**
 * Dado o head de uma linked list e uma valor x, particione de forma que todos os
 * nodes menores que x venham antes dos nodes maiores e iguais a x.
 *
 * Você deve preservar a ordem relativa original em ambas as partições.
 *
 * Restrições:
 *
 * Restrições:
 *
 * O números de nodes na lista está no intervalo [0, 200].
 *
 * -100 <= Node.val <= 100
 *
 * -200 <= x <= 200
 */
const partition = function (head, x) {
  let lessStart = null;
  let lessEnd = null;
  let greaterEqualStart = null;
  let greaterEqualEnd = null;
  while (head) {
    if (head.val < x) {
      if (!lessStart) {
        lessStart = head;
        lessEnd = head;
      } else {
        lessEnd.next = head;
        lessEnd = head;
      }
    } else if (!greaterEqualStart) {
      greaterEqualStart = head;
      greaterEqualEnd = head;
    } else {
      greaterEqualEnd.next = head;
      greaterEqualEnd = head;
    }
    head = head.next;
  }
  if (greaterEqualEnd) greaterEqualEnd.next = null;
  if (!lessEnd) {
    return greaterEqualStart;
  }
  lessEnd.next = greaterEqualStart;
  return lessStart;
};

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

function generateListNode(values) {
  const dummy = new ListNode(-1);
  let node = dummy;

  for (const val of values) {
    node.next = new ListNode(val);
    node = node.next;
  }
  return dummy.next;
}
